Framework/Indexing/src/IndexInfo.cpp

```diff
@@ -0,0 +1,122 @@
+}
+
+/// Copy constructor.
+IndexInfo::IndexInfo(const IndexInfo &other) {
+  if (other.m_isLegacy) {
+    // Workaround while IndexInfo is not holding index data stored in
+    // MatrixWorkspace: build IndexInfo based on data in ISpectrum.
+    auto size = other.m_getSize();
+    auto &specNums = m_spectrumNumbers.access();
+    auto &detIDs = m_detectorIDs.access();
+    for (size_t i = 0; i < size; ++i) {
+      specNums.push_back(other.m_getSpectrumNumber(i));
+      const auto &set = other.m_getDetectorIDs(i);
+      detIDs.emplace_back(set.begin(), set.end());
+    }
+  } else {
+    m_spectrumNumbers = other.m_spectrumNumbers;
+    m_detectorIDs = other.m_detectorIDs;
+  }
+}
+
+/// The *local* size, i.e., the number of spectra in this partition.
+size_t IndexInfo::size() const {
+  if (m_isLegacy)
+    return m_getSize();
+  return m_spectrumNumbers->size();
+}
+
+/// Returns the spectrum number for given index.
+specnum_t IndexInfo::spectrumNumber(const size_t index) const {
+  if (m_isLegacy)
+    return m_getSpectrumNumber(index);
+  return (*m_spectrumNumbers)[index];
+}
+
+/// Return a vector of the detector IDs for given index.
+std::vector<detid_t> IndexInfo::detectorIDs(const size_t index) const {
+  if (m_isLegacy) {
+    const auto &ids = m_getDetectorIDs(index);
+    return std::vector<detid_t>(ids.begin(), ids.end());
+  }
+  return (*m_detectorIDs)[index];
+}
+
+/// Set a spectrum number for each index.
+void IndexInfo::setSpectrumNumbers(std::vector<specnum_t> &&spectrumNumbers) & {
+  // No test of m_isLegacy, we cannot have non-const access in that case.
+  if (m_spectrumNumbers->size() != spectrumNumbers.size())
+    throw std::runtime_error(
+        "IndexInfo: Size mismatch when setting new spectrum numbers");
+  m_spectrumNumbers.access() = std::move(spectrumNumbers);
+}
+
+/// Set a single detector ID for each index.
+void IndexInfo::setDetectorIDs(const std::vector<detid_t> &detectorIDs) & {
+  // No test of m_isLegacy, we cannot have non-const access in that case.
+  if (m_detectorIDs->size() != detectorIDs.size())
+    throw std::runtime_error(
+        "IndexInfo: Size mismatch when setting new detector IDs");
+
+  auto &detIDs = m_detectorIDs.access();
+  for (size_t i = 0; i < detectorIDs.size(); ++i)
+    detIDs[i] = {detectorIDs[i]};
+}
+
+/// Set a vector of detector IDs for each index.
+void IndexInfo::setDetectorIDs(
+    // No test of m_isLegacy, we cannot have non-const access in that case.
+    std::vector<std::vector<detid_t>> &&detectorIDs) & {
+  if (m_detectorIDs->size() != detectorIDs.size())
+    throw std::runtime_error(
+        "IndexInfo: Size mismatch when setting new detector IDs");
+
+  auto &detIDs = m_detectorIDs.access();
+  detIDs = std::move(detectorIDs);
+  for (auto &ids : detIDs) {
+    std::sort(ids.begin(), ids.end());
+    ids.erase(std::unique(ids.begin(), ids.end()), ids.end());
+  }
+}
+
+} // namespace Indexing
+} // namespace Mantid
```

Comment:
This looks solid to me. I guess partitioning is something that can be added later?